docs/multi_process.md

```diff
@@ -1,176 +1,568 @@
-8
-# The psum is performed over all mapped devices across the pod slice
->>> xs = jax.numpy.ones(jax.local_device_count())
->>> jax.pmap(lambda x: jax.lax.psum(x, 'i'), axis_name='i')(xs)
-ShardedDeviceArray([32., 32., 32., 32., 32., 32., 32., 32.], dtype=float32)
-```
-
-**It’s very important that all processes run the same cross-process computations
-in the same order.** Running the same JAX Python program in each process is
-usually sufficient. Some common pitfalls to look out for that may cause
-differently-ordered computations despite running the same program:
-
-*   Processes passing differently-shaped inputs to the same parallel function
-    can cause hangs or incorrect return values. Differently-shaped inputs are
-    safe so long as they result in identically-shaped per-device data shards
-    across processes; e.g. passing in different leading batch sizes in order to
-    run on different numbers of local devices per process is ok, but having each
-    process pad its batch to a different max example length is not.
-
-*   “Last batch” issues where a parallel function is called in a (training)
-    loop, and one or more processes exit the loop earlier than the rest. This
-    will cause the rest to hang waiting for the already-finished processes to
-    start the computation.
-
-*   Conditions based on non-deterministic ordering of collections can cause code
-    processes to hang. For example, iterating over
-    `set` on current Python versions or `dict` [before Python 3.7](https://mail.python.org/pipermail/python-dev/2017-December/151283.html)
-    may result in a different ordering on different processes, even with the
-    same insertion order.
+# target (micro)batch size across the whole cluster
+batch_size = 1024
+# how many examples each process should load per batch
+per_process_batch_size = batch_size // jax.process_count()
+# how many examples each device will process per batch
+per_device_batch_size = batch_size // jax.device_count()
+
+# make a data-parallel mesh and sharding
+mesh = jax.make_mesh((jax.device_count(),), ('batch'))
+sharding = NamedSharding(mesh, P('batch'))
+
+# our "data loader". each process loads a different set of "examples".
+process_batch = np.random.rand(per_process_batch_size, 2048, 42)
+
+# assemble a global array containing the per-process batches from all processes
+global_batch = jax.make_array_from_process_local_data(sharding, process_batch)
+
+# sanity check that everything got sharded correctly
+assert global_batch.shape[0] == batch_size
+assert process_batch.shape[0] == per_process_batch_size
+assert global_batch.addressable_shards[0].data.shape[0] == per_device_batch_size
+```
+
+{func}`jax.make_array_from_single_device_arrays` is the most general way to
+build a process-spanning array. It's often used after performing
+{func}`jax.device_put`s to send to each device its required data. It is the most
+low-level, since all data movement is performed manually (via e.g.
```

Comment:
```suggestion
{func}`jax.device_put`s to send each device its required data. This is the
lowest-level option, since all data movement is performed manually (via e.g.
```